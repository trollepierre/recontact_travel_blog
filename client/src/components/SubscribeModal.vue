<template>
  <div class="subscribe-modal-wrapper">
    <modal class="subscribe-modal" name="subscribe-modal" @before-open="beforeOpen" :height="415">

      <!-- modal header-->
      <div class="subscribe-modal__header">
        <h2 class="subscribe-modal__title">S'abonner</h2>
      </div>

      <!-- modal body -->
      <div class="subscribe-modal__body">
        <form class="subscribe-modal__form">

          <p class="subscribe-modal__error" v-if="error" aria-live="polite">{{error}}</p>

          <label class="subscribe-modal__label" for="subscribe-content">Email :</label>
          <input class="subscribe-modal__email" id="subscribe-content" v-model="email"/>
        </form>
      </div>

      <!-- modal footer -->
      <div class="subscribe-modal__footer">
        <div class="subscribe-modal__actions">
          <button class="subscribe-modal__action subscribe-modal__action--send" @click="sendSubscription">Envoyer</button>
          <button class="subscribe-modal__action subscribe-modal__action--cancel" @click="cancelSubscription">Annuler</button>
        </div>
      </div>

    </modal>
  </div>
</template>

<script>
  import subscriptionsApi from '@/api/subscriptions';
  import notificationService from '@/services/notification';

  export default {
    name: 'SubscribeModal',
    data() {
      return {
        email: null,
        error: null,
      };
    },
    methods: {
      beforeOpen() {
        this._resetFeedback();
        this._removeError();
      },

      sendSubscription() {
        this._removeError();
        if (!this.email || this.email.trim().length === 0) {
          // todo check email pattern
          this.error = 'Vous devez saisir un email.';
          return;
        }

        subscriptionsApi.subscribe(this.email)
          .then(this.displaySuccessNotification)
          .then(this._closeModal)
          .catch(() => {
            this.error = 'Erreur lors de la prise en compte de ton abonnement.';
          });
      },

      displaySuccessNotification() {
        const message = 'Ton abonnement aux alertes de Recontact Me a été pris en compte.';
        notificationService.success(this, message);
      },

      cancelSubscription() {
        this._closeModal();
      },

      _resetFeedback() {
        this.subscribe = null;
      },

      _removeError() {
        this.error = null;
      },

      _closeModal() {
        this.$modal.hide('subscribe-modal');
      },
    },
  };

</script>

<style scoped>
  .subscribe-modal__header {
    background-color: #eef0f4;
    padding: 10px 20px;
    justify-content: center;
    display: flex;  }

  .subscribe-modal__body {
    padding: 25px 20px;
    background: #fff;
    height: 216px;
  }

  .subscribe-modal__form {
    padding: 0;
    margin: 0;
  }

  .subscribe-modal__title {
    margin: 0;
    font-size: 24px;
    color: #333333;
    height: 40px;
    line-height: 40px;
  }

  .subscribe-modal__error {
    background-color: #ea6344;
    color: white;
    border-radius: 3px;
    padding: 10px;
    margin: 0 0 20px;
  }

  .subscribe-modal__label {
    display: block;
    margin-bottom: 10px;
  }

  .subscribe-modal__text {
    width: 100%;
    border: 1px solid #d8dde6;
    resize: none;
    overflow: auto;
    height: 152px;
    font-size: 16px;
    box-sizing: border-box;
    padding: 5px;
  }

  .subscribe-modal__footer {
    padding: 20px;
    background: #fff;
    text-align: right;
  }

  .subscribe-modal__action {
    display: inline-block;
    padding: 4px 15px;
    font-size: 14px;
    line-height: 16px;
    color: #fff;
    text-align: center;
    vertical-align: middle;
    cursor: pointer;
    background: #008a00;
    outline: none;
    border: none;
    -webkit-border-radius: 3px;
    border-radius: 3px;
    height: 34px;
    font-weight: 500;
    margin: 0;
  }

  .subscribe-modal__action:focus {
    border: 3px #000000 solid;
  }

  .subscribe-modal__action:hover {
    opacity: .85;
  }

  .subscribe-modal__action--cancel {
    background: #fff;
    border: 1px solid #d8dde6;
    color: #333333;
  }

  .subscribe-modal__action--cancel:hover {
    box-shadow: 0 0 3px 0 #d8dde6;
  }
</style>
