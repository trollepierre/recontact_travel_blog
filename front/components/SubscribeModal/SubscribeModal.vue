<template>
  <div class="subscribe-modal-wrapper">
    <modal
      :width="widthModal"
      :height="315"
      class="subscribe-modal"
      name="subscribe-modal"
      @before-open="beforeOpen"
      @opened="opened">
      <div class="subscribe-modal__header">
        <h2 class="subscribe-modal__title">
          {{ $t("subscribe") }}
        </h2>
      </div>

      <div class="subscribe-modal__body">
        <form
          class="subscribe-modal__form"
          @submit="submit">
          <p
            v-if="error"
            class="subscribe-modal__error"
            aria-live="polite">
            {{ error }}
          </p>

          <p class="subscribe-modal__text">
            {{ $t("modalText") }}
          </p>

          <label
            class="subscribe-modal__label"
            for="subscribe-content">
            {{ $t("email") }}
          </label>
          <input
            id="subscribe-content"
            v-model="email"
            class="subscribe-modal__email"
            placeholder="pierre@recontact.me">
        </form>
      </div>

      <div class="subscribe-modal__footer">
        <div class="subscribe-modal__actions">
          <button
            class="subscribe-modal__action subscribe-modal__action--send"
            @click="sendSubscription">
            {{ $t("confirm") }}
          </button>
          <button
            class="subscribe-modal__action subscribe-modal__action--cancel"
            @click="cancelSubscription">
            {{ $t("cancel") }}
          </button>
        </div>
      </div>
    </modal>
  </div>
</template>

<script>
  import notificationsService from '../../services/services/notifications'
  import subscriptionsApi from '../../services/api/subscriptions'
  import { screenWidth, PHONE_PORTRAIT_TO_LANDSCAPE } from '../../services/utils/screen/screen-utils'

  export default {
    name: 'SubscribeModal',
    data() {
      return {
        email: null,
        error: null,
      }
    },
    computed: {
      widthModal() {
        return screenWidth < PHONE_PORTRAIT_TO_LANDSCAPE ? 300 : undefined
      },
    },
    methods: {
      beforeOpen() {
        this._resetEmail()
        this._removeError()
      },

      opened() {
        this.trackEvent()
        this._focusOnInput()
        this._closeOnEscapeKey()
      },

      _closeOnEscapeKey() {
        document.addEventListener('keydown', e => {
          if (e.keyCode === 27) {
            this._closeModal()
          }
        })
      },

      _focusOnInput() {
        this.$el.querySelector('input#subscribe-content').focus()
      },

      sendSubscription() {
        this._removeError()
        /* eslint-disable no-useless-escape */
        const regex = new RegExp('^[_A-Za-z0-9-\+-]+(\.[_A-Za-z0-9-\+-]+)*@[A-Za-z0-9-]+(\.[A-Za-z0-9-\+-]+)*(\.[A-Za-z]{2,})$')
        if (!regex.exec(this.email)) {
          this.error = this.$t('emailError')
          return
        }

        subscriptionsApi.subscribe(this.email)
          .then(this.displaySuccessNotification)
          .then(this._closeModal)
          .catch(() => {
            this.error = this.$t('subscriptionError')
          })
      },

      displaySuccessNotification() {
        notificationsService.success(this, this.$t('subscriptionSuccess'))
      },

      cancelSubscription() {
        this._closeModal()
      },

      _resetEmail() {
        this.email = null
      },

      _removeError() {
        this.error = null
      },

      _closeModal() {
        this.$modal.hide('subscribe-modal')
      },

      submit(e) {
        e.preventDefault()
        this.sendSubscription()
      },

      trackEvent() {
        this.$ga.event({
          eventCategory: 'Subscribe Modal',
          eventAction: 'open',
          eventLabel: 'Subscribe modal is opened',
        })
      },
    },
    i18n: {
      messages: {
        fr: {
          subscribe: 'S’abonner',
          modalText: 'Recevoir un email à chaque nouvel article du voyage  !',
          email: 'Email  :',
          confirm: 'Confirmer',
          cancel: 'Annuler',
          emailError: 'Vous devez saisir un email valide. (ex. : nom@exemple.fr)',
          subscriptionError: 'Erreur lors de la prise en compte de ton abonnement.',
          subscriptionSuccess: 'Ton abonnement aux alertes de Recontact Me a été pris en compte.',
        },
        en: {
          subscribe: 'Subscribe',
          modalText: 'Receive one email for each new article.',
          email: 'Email:',
          confirm: 'Confirm',
          cancel: 'Cancel',
          emailError: 'Please insert a valid email (ex. : name@example.com)',
          subscriptionError: 'Error during the subscription.',
          subscriptionSuccess: 'Your subscription to Recontact Me alerts has been taken into consideration.',
        },
      },
    },
  }

</script>

<style scoped>
  .subscribe-modal__header {
    background-color: #eef0f4;
    padding: 10px 20px;
    justify-content: center;
    display: flex;
  }

  .subscribe-modal__body {
    padding: 5px 20px;
    background: #fff;
    height: 125px;
  }

  .subscribe-modal__form {
    padding: 0;
    margin: 0;
  }

  .subscribe-modal__title {
    margin: 0;
    font-size: 24px;
    color: #333333;
    height: 40px;
    line-height: 40px;
  }

  .subscribe-modal__error {
    background-color: #ea6344;
    color: white;
    border-radius: 3px;
    padding: 10px;
    margin: 0;
  }

  .subscribe-modal__label {
    display: block;
    margin-bottom: 10px;
  }

  .subscribe-modal__email {
    width: 250px;
    font-size: 16px;
    border-color: #424547;
  }

  .subscribe-modal__text {
    width: 100%;
    resize: none;
    overflow: auto;
    font-size: 16px;
    box-sizing: border-box;
    padding: 5px;
  }

  .subscribe-modal__footer {
    padding-right: 20px;
    margin-top: 75px;
    background: #fff;
    text-align: right;
  }

  .subscribe-modal__action {
    display: inline-block;
    padding: 4px 15px;
    font-size: 14px;
    line-height: 16px;
    color: #fff;
    text-align: center;
    vertical-align: middle;
    cursor: pointer;
    background: #008a00;
    outline: none;
    border: none;
    -webkit-border-radius: 3px;
    border-radius: 3px;
    height: 34px;
    font-weight: 500;
    margin: 0;
  }

  .subscribe-modal__action:focus {
    border: 3px #000000 solid;
  }

  .subscribe-modal__action:hover {
    opacity: .85;
  }

  .subscribe-modal__action--cancel {
    background: #fff;
    border: 1px solid #d8dde6;
    color: #333333;
  }

  .subscribe-modal__action--cancel:hover {
    box-shadow: 0 0 3px 0 #d8dde6;
  }
</style>
