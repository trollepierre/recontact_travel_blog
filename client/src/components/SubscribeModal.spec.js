import Vue from 'vue';
import SubscribeModal from './SubscribeModal';
import subscriptionsApi from '../api/subscriptions';
import notificationsService from '../services/notifications';

xdescribe('Component | SubscribeModal.vue', () => {
  let wrapper
  const email = 'pierre@recontact.me';

  beforeEach(() => {

    const Constructor = Vue.extend(SubscribeModal);


    let localVue; localVue = createLocalVue(); wrapper = shallowMount(AppHeader, { localVue,
      data() {
        return {
          email,
        };
      },
    })
  });

  it('should be named "SubscribeModal"', () => {
    expect(wrapper.name()).toEqual('SubscribeModal');
  });

  describe('template', () => {
    it('should match snapshot', () => {
      let localVue; localVue = createLocalVue(); wrapper = shallowMount(AppHeader, { localVue })

      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('rendering', () => {
    it('should display the modal', () => {
      component.$modal.show('subscribe-modal');

      return Vue.nextTick().then(() => {
        expect(wrapper.find('.subscribe-modal')).to.exist;
      });
    });
  });

  describe('#beforeOpen', () => {
    it('should reset email', () => {

      wrapper.vm.email = 'Coucou@contact.me';


      component.beforeOpen();


      expect(wrapper.vm.email).toEqual(null);
    });

    it('should remove error', () => {

      wrapper.vm.error = 'C\'est un problème';


      component.beforeOpen();


      expect(wrapper.vm.error).toEqual(null);
    });
  });

  describe('#opened', () => {
    beforeEach(() => {
      sinon.stub(component, '_focusOnInput');
      sinon.stub(component, '_closeModal');
    });

    afterEach(() => {
      component._focusOnInput.restore();
      component._closeModal.restore();
    });

    it('should focusOnInput', () => {

      component.opened();


      expect(component._focusOnInput).toHaveBeenCalledWith();
    });

    it('should close on escape key', () => {

      component.opened();


      const e = document.createEvent('Events');
      e.initEvent('keydown', true, true);
      e.keyCode = 27;
      document.dispatchEvent(e);

      return Vue.nextTick().then(() => {
        expect(component._closeModal).toHaveBeenCalledWith();
      });
    });

    it('should not close on any other key than space', () => {

      component.opened();


      const e = document.createEvent('Events');
      e.initEvent('keydown', true, true);
      e.keyCode = 13;
      document.dispatchEvent(e);

      return Vue.nextTick().then(() => {
        expect(component._closeModal).not.toHaveBeenCalledWith();
      });
    });
  });

  describe('#_focusOnInput', () => {
    it.skip('should focus on input subscribe content', (done) => {

      component.$modal.show('subscribe-modal');


      setTimeout(() => {

        const inputSubscribe = wrapper.find('input#subscribe-content');
        expect(inputSubscribe).to.have.focus();
        done();
      }, 100);
    });
  });

  describe('#sendSubscription', () => {
    beforeEach(() => {
      sinon.stub(subscriptionsApi, 'subscribe').resolves();
      sinon.stub(notificationsService, 'success');
    });

    afterEach(() => {
      subscriptionsApi.subscribe.restore();
      notificationsService.success.restore();
    });


    it('should remove error', () => {

      wrapper.vm.error = 'C\'est un problème';


      component.sendSubscription();


      expect(wrapper.vm.error).toEqual(null);
    });

    describe('when email is empty', () => {
      it('should set error', () => {

        wrapper.vm.email = '';


        component.sendSubscription();


        expect(wrapper.vm.error).toEqual('emailError');
      });

      it('should not call sendSubscription', () => {

        wrapper.vm.email = '';


        component.sendSubscription();


        expect(subscriptionsApi.subscribe).not.toHaveBeenCalled;
      });
    });

    describe('when email does not follow good pattern', () => {
      it('should set error', () => {

        wrapper.vm.email = 'pierrerecontact';


        component.sendSubscription();


        expect(wrapper.vm.error).toEqual('emailError');
      });

      it('should not call sendSubscription', () => {

        wrapper.vm.email = 'pierre@recontact.m';


        component.sendSubscription();


        expect(subscriptionsApi.subscribe).not.toHaveBeenCalled;
      });
    });

    it('should call the API with good params', () => {

      component.sendSubscription(email);


      expect(subscriptionsApi.subscribe).toHaveBeenCalledWith(email);
    });

    it('should display success notification', () => {

      notificationsService.success.resolves({});


      component.sendSubscription();


      return Vue.nextTick().then(() => {
        const message = 'subscriptionSuccess';
        expect(notificationsService.success).toHaveBeenCalledWith(component, message);
      });
    });

    it('should close modal', () => {

      component.$modal.show('subscribe-modal');
      wrapper.vm.email = 'email@recontact.me';


      component.sendSubscription();


      return Vue.nextTick().then(() => {
        expect(wrapper.find('.subscribe-modal')).not.to.exist;
      });
    });

    describe('when sendSubscription fails', () => {
      beforeEach(() => {
        subscriptionsApi.subscribe.restore();
        component.$modal.show('subscribe-modal');
        wrapper.vm.email = 'email@recontact.me';
        sinon.stub(subscriptionsApi, 'subscribe').rejects(new Error('e'));
      });

      it('should not close modal', () => {

        component.sendSubscription();


        return Vue.nextTick().then(() => {
          expect(wrapper.find('.subscribe-modal')).to.exist;
        });
      });

      it('should set error', () => {

        component.sendSubscription();


        return Vue.nextTick().then(() => {
          expect(wrapper.vm.error).toEqual('subscriptionError');
        });
      });
    });
  });

  describe('#cancelSubscription', () => {
    it('should close modal', () => {

      component.$modal.show('subscribe-modal');


      component.cancelSubscription();


      return Vue.nextTick().then(() => {
        expect(wrapper.find('.subscribe-modal')).not.to.exist;
      });
    });
  });

  it.skip('should sendSubscription on click on "send" button', () => {

    component.$modal.show('subscribe-modal');
    sinon.stub(component, 'sendSubscription');

    return Vue.nextTick().then(() => {
      const myButton = wrapper.find('.subscribe-modal__action--send');


      myButton.click();


      expect(component.sendSubscription).toHaveBeenCalled;
    });
  });

  it.skip('should cancelSubscription on click on "cancel" button', () => {

    component.$modal.show('subscribe-modal');
    sinon.stub(component, 'cancelSubscription');

    return Vue.nextTick().then(() => {
      const myButton = wrapper.find('.subscribe-modal__action--cancel');


      myButton.click();


      expect(component.cancelSubscription).toHaveBeenCalled;
    });
  });

  describe('locales', () => {
    const languages = Object.keys(SubscribeModal.i18n.messages);

    it('contains 2 languages', () => {
      expect(languages.length).toEqual(2);
      expect(languages).toEqual(['fr', 'en']);
    });

    describe('each language', () => {
      describe('fr', () => {
        const locales = Object.keys(SubscribeModal.i18n.messages.fr);

        it('contains 8 locales', () => {
          expect(locales.length).toEqual(8);
          expect(locales).toEqual([
            'subscribe',
            'modalText',
            'email',
            'confirm',
            'cancel',
            'emailError',
            'subscriptionError',
            'subscriptionSuccess',
          ]);
        });
      });

      describe('en', () => {
        const locales = Object.keys(SubscribeModal.i18n.messages.en);

        it('contains 8 locales', () => {
          expect(locales.length).toEqual(8);
          expect(locales).toEqual([
            'subscribe',
            'modalText',
            'email',
            'confirm',
            'cancel',
            'emailError',
            'subscriptionError',
            'subscriptionSuccess',
          ]);
        });
      });
    });
  });
});
