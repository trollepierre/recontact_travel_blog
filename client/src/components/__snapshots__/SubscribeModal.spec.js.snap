// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | SubscribeModal.vue template should match snapshot 1`] = `
<div
  class="subscribe-modal-wrapper"
>
  <modal
    class="subscribe-modal"
    height="315"
    name="subscribe-modal"
  >
    <div
      class="subscribe-modal__header"
    >
      <h2
        class="subscribe-modal__title"
      >
        
        subscribe
      
      </h2>
    </div>
     
    <div
      class="subscribe-modal__body"
    >
      <form
        class="subscribe-modal__form"
      >
        <!---->
         
        <p
          class="subscribe-modal__text"
        >
          
          modalText
        
        </p>
         
        <label
          class="subscribe-modal__label"
          for="subscribe-content"
        >
          
          email
        
        </label>
         
        <input
          class="subscribe-modal__email"
          id="subscribe-content"
          placeholder="pierre@recontact.me"
        />
      </form>
    </div>
     
    <div
      class="subscribe-modal__footer"
    >
      <div
        class="subscribe-modal__actions"
      >
        <button
          class="subscribe-modal__action subscribe-modal__action--send"
        >
          
          confirm
        
        </button>
         
        <button
          class="subscribe-modal__action subscribe-modal__action--cancel"
        >
          
          cancel
        
        </button>
      </div>
    </div>
  </modal>
</div>
`;
